Nightly test refactor with Artifactory support

jdavies-st opened the pull request:
This is a fork of @stsci-hack's PR #2529 that we are using for testing on Jenkins.

Test results here
https://boyle.stsci.edu:8081/job/RT/job/JWST-jdavies/

Closes #2529.

jdavies-st commented:
Note to self.  New reg tests have been added in #2676.  Do not forget about them here.

pllim commented:
It passed :scream: :clap:

jdavies-st commented:
Yup!  I notice that we have 3 skips to the 2 skips on `master`, but there's an extra `xfail` on master.  So perhaps one more change.

Current test results on this PR:
https://boyle.stsci.edu:8081/job/RT/job/JWST-jdavies/

pllim commented:
340 warnings though... :stuck_out_tongue_closed_eyes:

stscicrawford commented:
Looks like the Jenkins test needs to be restarted

jdavies-st commented:
So I think this is ready for review. As you can see, most of the work in this PR is done by @stsci-hack, with some Artifactory and CI-related changes that I've added.

The regression tests ran this morning and all passed:

https://boyle.stsci.edu:8081/job/RT/job/JWST-jdavies/

So hopefully this is ready to merge?

I assume merging into master, i.e. for 0.12.1 patch is OK?

stscicrawford commented:
As mentioned and separate from this PR, there should be documentation on how to run the tests, update the tests, and ok-ify the truth files.   

Also, it looks like our current regression tests are failing with merging of Jane's and Howard's PR's yesterdays.   Was it just the timing or is there some reason this is not picking up those failures?

jdavies-st commented:
This PR does not contain those changes from master yesterday.  It has not been rebased to master for a couple days, hence the difference.

pllim commented on jwst/tests/base_test.py:
```diff
@@ -0,0 +1,153 @@
+            self.docopy = True
+
+    def raw_from_asn(self, asn_file):
+        return raw_from_asn(asn_file)
```
Why wrap this in a method?

pllim commented on jwst/tests/base_test.py:
```diff
@@ -0,0 +1,153 @@
+
+
+# function(s) for tests written using functional form, not as BaseTest class tests
+def raw_from_asn(asn_file):
```
There is already such functionality in `ci-watson` for HST. Are you able to re-use that for JWST?

jdavies-st replied:
JWST association files are different I believe.

pllim replied:
Perhaps a future PR but I think it might benefit more people if JWST's version of `raw_from_asn` is moved to `ci-watson`. Such functionality is useful beyond `jwst` pipeline testing. Actually, maybe for both HST and JWST, they are useful beyond `ci-watson` but I am hesitant to suggest `stsci.tools` without giving it more thought.

pllim commented on jwst/tests/base_test.py:
```diff
@@ -0,0 +1,153 @@
+            for funcargs in funcarglist], ids=idlist)
+
+
+@pytest.mark.usefixtures('_jail')
```
This is already inherited, so I don't think you need to re-declare `_jail` here.

pllim commented on jwst/tests_nightly/general/fgs/test_fgs_image2_1.py:
```diff
@@ -0,0 +1,27 @@
+    input_loc = 'fgs'
+    ref_loc = ['test_image2pipeline', 'truth']
+
+    def test_fgs_image2pipeline1(self, _jail):
```
This is already inherited, so I don't think you need to re-declare `_jail` here. This comment also applies to all similar occurrences.

pllim commented on jwst/tests_nightly/general/framework/test_self.py:
```diff
@@ -1,18 +1,12 @@
-
-
-def test_bigdata(_bigdata):
+@pytest.mark.bigdata
```
I feel like this is overkill. This whole file can be deleted.

jdavies-st replied:
Good catch.

jdavies-st commented on jwst/conftest.py:
```diff
@@ -36,24 +36,6 @@ def check_url(url):
-
-
 @pytest.fixture
 def mk_tmp_dirs():
```
`mk_tmp_dirs` is used in some of our `--slow --bigdata` tests in `associations` and `pipeline` which don't get run in any automated way yet, i.e. not on our per-PR CI, nor during our nightly tests.  I feel this is a project for a different PR.  Agree we should move to using `_jail`.

Yes, I think `check_url` and `BigdataError` can be removed.

jdavies-st commented on jwst/tests/base_classes.py:
```diff
@@ -0,0 +1,167 @@
+        other tests.
+        """
+        # create working directory specified for the test
+        if not tmpdir.ensure(self.subdir, dir=True):
```
Yes!

jdavies-st commented:
@pllim I think I've addressed your comments in the last several commits.

The only thing I have not done yet is sort out the uploading of the failed test result files to Artifactory.  The `// Configure artifactory ingest` above uploads the build information to artifactory as a minimum.  I may need some help on the final piece of the puzzle.

pllim commented:
Great work, James and Warren!
